server/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java

```diff
@@ -202,4 +202,75 @@ public Releasable dropHeadersContext() {
+     * producing a cluster state after each task. This allows executing the tasks in the batch as a
+     * series of executions, each taking an input cluster state and producing a new cluster state that serves as the
+     * input of the next task in the batch.
+     */
+    abstract class DefaultBatchExecutor<T extends ClusterStateTaskListener> implements ClusterStateTaskExecutor<T> {
+
+        /**
+         * Executes the provided task from the batch.
+         *
+         * @param taskContext The task to be executed.
+         * @param curState The cluster state on which the task should be executed.
+         * @return The resulting cluster state after executing this task.
+         */
+        public abstract ClusterState executeTask(TaskContext<T> taskContext, ClusterState curState);
+
+        /**
+         * Called once all tasks in the batch have finished execution. It should return a cluster state that reflects
+         * the execution of all the tasks.
+         *
+         * @param batchExecutionContext The tasks that were executed.
+         * @param initState The cluster state before executing the tasks.
+         * @param curState The cluster state resulting from the execution of all the tasks.
+         * @return The resulting cluster state after executing all the tasks.
+         */
+        public ClusterState afterBatchExecution(
+            BatchExecutionContext<T> batchExecutionContext,
+            ClusterState initState,
+            ClusterState curState
+        ) {
+            return curState;
+        }
+
+        /**
+         * Called if executing a task in the batch finished successfully, and before the execution of the next
+         * task in the batch.
+         *
+         * @param taskContext The task that successfully finished execution.
+         */
+        public void taskSucceeded(TaskContext<T> taskContext) {
+            taskContext.success(() -> {});
+        }
+
+        /**
+         * Called if executing a task in the batch fails by throwing an Exception, and before the execution of the
+         * next task in the batch.
+         *
+         * @param taskContext The task that failed with an exception.
+         * @param e The Exception thrown by the task execution.
+         */
+        public void taskFailed(TaskContext<T> taskContext, Exception e) {
+            taskContext.onFailure(e);
+        }
+
+        @Override
+        public final ClusterState execute(BatchExecutionContext<T> batchExecutionContext) throws Exception {
+            var initState = batchExecutionContext.initialState();
+            var curState = initState;
+            for (final var taskContext : batchExecutionContext.taskContexts()) {
+                try (var ignored = taskContext.captureResponseHeaders()) {
+                    curState = executeTask(taskContext, curState);
+                    taskSucceeded(taskContext);
```

Comment:
Rather than asking implementations to manipulate the `TaskContext` themselves, I think they will almost always just want to call a method on the underlying task. Can we do this instead?

```suggestion
                    taskContext.success(() -> taskSucceeded(taskContext.getTask()));
```

Reply:
What about when the task is itself a `ClusterStateAckListener`? In that case we'd have to ask the executor to implement an empty `taskSucceeded` and in the batch execution inspect the type of the task?  Or `taskSucceeded` would optionally return the task? Both seem not much better. Or maybe there is a simpler way to deal with that? Or do we explicitly not want to accommodate that case?

Reply:
I think we probably want a whole different executor for ack-sensitive tasks, with different callbacks. Typically these things complete their listener in `onAllNodesAcked()` rather than when the publication completes.

Reply:
So I guess I'll just drop it from this PR.

Reply:
:+1: sounds good to me